socketio/src/asynchronous/client/builder.rs

```diff
@@ -0,0 +1,329 @@
+            Some(ref mut map) => {
+                map.insert(key.into(), val.into());
+            }
+            None => {
+                let mut map = HeaderMap::default();
+                map.insert(key.into(), val.into());
+                self.opening_headers = Some(map);
+            }
+        }
+        self
+    }
+
+    /// Specifies which EngineIO [`TransportType`] to use.
+    ///
+    /// # Example
+    /// ```rust
+    /// use rust_socketio::{asynchronous::ClientBuilder, TransportType};
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let socket = ClientBuilder::new("http://localhost:4200/")
+    ///         // Use websockets to handshake and connect.
+    ///         .transport_type(TransportType::Websocket)
+    ///         .connect()
+    ///         .await
+    ///         .expect("connection failed");
+    /// }
+    /// ```
+    pub fn transport_type(mut self, transport_type: TransportType) -> Self {
+        self.transport_type = transport_type;
+
+        self
+    }
+
+    /// Connects the socket to a certain endpoint. This returns a connected
+    /// [`Client`] instance. This method returns an [`std::result::Result::Err`]
+    /// value if something goes wrong during connection. Also starts a separate
+    /// thread to start polling for packets. Used with callbacks.
+    /// # Example
+    /// ```rust
+    /// use rust_socketio::{asynchronous::ClientBuilder, Payload};
+    /// use serde_json::json;
+    /// use futures_util::future::FutureExt;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let mut socket = ClientBuilder::new("http://localhost:4200/")
+    ///         .namespace("/admin")
+    ///         .on("error", |err, _| async move { eprintln!("Error: {:#?}", err) }.boxed())
+    ///         .connect()
+    ///         .await
+    ///         .expect("connection failed");
+    ///
+    ///     // use the socket
+    ///     let json_payload = json!({"token": 123});
+    ///
+    ///     let result = socket.emit("foo", json_payload).await;
+    ///
+    ///     assert!(result.is_ok());
+    /// }
+    /// ```
+    pub async fn connect(self) -> Result<Client> {
+        let socket = self.connect_manual().await?;
+        let mut socket_clone = socket.clone();
+
+        // Use thread to consume items in iterator in order to call callbacks
+        tokio::runtime::Handle::current().spawn(async move {
+            loop {
+                // tries to restart a poll cycle whenever a 'normal' error occurs,
+                // it just panics on network errors, in case the poll cycle returned
+                // `Result::Ok`, the server receives a close frame so it's safe to
+                // terminate
+                for packet in socket_clone.next().await {
+                    if let e @ Err(Error::IncompleteResponseFromEngineIo(_)) = packet {
+                        //TODO: 0.3.X handle errors
+                        panic!("{}", e.unwrap_err());
```

Comment:
This panic the entire application actually... I know it's annoying but I don't really know what to do here. If the underlying transport is in an irreparable state, all future messages / actions are invalid. As this runs in the background I can't really report the issue to the user (or only if this method returns a channel, which is also a weird interface...)

Reply:
blocker: panicking libraries is a nightmare, especially one such as this where I (as the caller) can't prevent that situation from occurring.

IMO: log (other libraries such as actix log at the trace level) + throw an error packet

Reply:
Oke, I added the tracing. The error also shows up in case the user polls the stream. I don't feel good about it as this might lead to weird situations, but you are right about the panic.